Porting of space filling curves from "Neo4j Spatial"

OliviaYtterbrink opened the pull request:
This is a separate module so that it can be used for future versions of the Neo4j Spatial library from which it was derived, so we don't maintain two copies.
The included Envelope and EnvelopTest are trimmed down from the Neo4j Spatial versions, so we might want to refactor that later, either include all, or have Neo4j Spatial extend this one.

Lojjs committed:
Porting of space filling curves from "Neo4j Spatial"

This is a separate module so that it can be used for future versions of the Neo4j Spatial library from which it was derived, so we don't maintain two copies.
The included Envelope and EnvelopTest are trimmed down from the Neo4j Spatial versions, so we might want to refactor that later, either include all, or have Neo4j Spatial extend this one.

craigtaverner committed:
Checkstyle issues on spatial index

craigtaverner committed:
Fixed licenses and notice for new neo4j-spatial-index module

OliviaYtterbrink committed:
Fix notice with correct year

sherfert commented on community/spatial-index/src/main/java/org/neo4j/gis/spatial/index/Envelope.java:
```diff
@@ -0,0 +1,389 @@
+            boolean result = true;
+            for ( int i = 0; i < min.length; i++ )
+            {
+                result = result && other.min[i] <= max[i] && other.max[i] >= min[i];
```
I see we have a shortcut in `covers`. If `ans` becomes false, we return it immediately. Can't we do the same here?

sherfert commented on community/spatial-index/src/main/java/org/neo4j/gis/spatial/index/Envelope.java:
```diff
@@ -0,0 +1,389 @@
+        }
+    }
+
+    // Attributes
```
This seems to be the only class (of this PR) that has Attributes at the bottom. I don't like that personally, but we should try to make it consistent at least.

sherfert commented on community/spatial-index/src/main/java/org/neo4j/gis/spatial/index/Envelope.java:
```diff
@@ -0,0 +1,389 @@
+        {
+            for ( int i = 0; i < min.length; i++ )
+            {
+                if ( other.min[i] < min[i] )
```
I think this will hit an ArrayIndexOutOfBoundsException if `other.dimension != this.dimension`. In other methods, this is checked beforehand. Any reason for not doing it here?

sherfert commented on community/spatial-index/src/main/java/org/neo4j/gis/spatial/index/Envelope.java:
```diff
@@ -0,0 +1,389 @@
+    {
+        this.min = min.clone();
+        this.max = max.clone();
+        if ( !isValid() )
```
`isValid` should be final, since it is called from a Constructor.

As we're doing this here and throwing an Exception if not valid, we will never have invalid Envelopes. This means we could removed the isValid checks in all the other methods. There are only two mutating methods, `expandToInclude` and `translate` and they both seem safe w.r.t `isValid`.

sherfert commented on community/spatial-index/src/main/java/org/neo4j/gis/spatial/index/curves/SpaceFillingCurve.java:
```diff
@@ -0,0 +1,483 @@
+            assert npointValues.length == length();
+        }
+
+        int length()
```
Should be final.

sherfert commented on community/spatial-index/src/main/java/org/neo4j/gis/spatial/index/curves/HilbertSpaceFillingCurve3D.java:
```diff
@@ -0,0 +1,248 @@
+            case 1:
+                return 'F'; // move up      000->001
+            case 2:
+                return 'U'; // move right   000->010
```
This and the comment above look wrong. Front and up?

sherfert reviewed (changes requested):
I have just minor comments, but still some that are worth going over. For the main part of the logic I will probably need to do another pass, since it is very complex.

OliviaYtterbrink committed:
Removed extra checks, cleaned up loops

isValid is only worth checking when constructing a new Envelope,
since all mutating methods will result in a valid Envelope.

OliviaYtterbrink committed:
Fixed comments

OliviaYtterbrink committed:
Cleaned up superclass

OliviaYtterbrink committed:
Use enum instead of strings as keys

OliviaYtterbrink committed:
Removed debug print

OliviaYtterbrink committed:
Checkstyle
